OpenRA.Mods.D2/Traits/Buildings/D2PlaceBuildingPreview.cs

```diff
@@ -0,0 +1,132 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The d2 mod Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.D2.Graphics;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.D2.Traits
+{
+	[Desc("Creates a d2 building placement preview.")]
+	public class D2PlaceBuildingPreviewInfo : FootprintPlaceBuildingPreviewInfo
+	{
+		[Desc("Speed of footprint animation.")]
+		public float Step = 0.04f;
+
+		protected override IPlaceBuildingPreview CreatePreview(WorldRenderer wr, ActorInfo ai, TypeDictionary init)
+		{
+			return new D2PlaceBuildingPreviewPreview(wr, ai, this, init);
+		}
+
+		public override object Create(ActorInitializer init)
+		{
+			return new D2PlaceBuildingPreview();
+		}
+	}
+
+	public class D2PlaceBuildingPreview { }
+
+	public class D2PlaceBuildingPreviewPreview : FootprintPlaceBuildingPreviewPreview
+	{
+		protected readonly D2PlaceBuildingPreviewInfo info;
+		protected readonly ActorInfo ai;
+		protected readonly D2BuildingInfo bi;
+		
+		protected readonly int cols;
+		protected readonly int rows;
+		protected readonly int halfWidth;
+		protected readonly int halfHeight;
+		protected readonly Rectangle bounds;
+
+		protected float t = 0.0f;
+	
+		public D2PlaceBuildingPreviewPreview(WorldRenderer wr, ActorInfo ai, D2PlaceBuildingPreviewInfo info, TypeDictionary init)
+			: base(wr, ai, info, init)
+		{
+			this.info = info;
+			this.ai = ai;
+
+			bi = ai.TraitInfoOrDefault<D2BuildingInfo>();
```

Comment:
You are not checking if `bi` is null, so it is better to just use `.TraitInfo<D2BuildingInfo>` here.

Reply:
Also added `Requires<D2BuildingInfo>`